pkg/controller/endpointslicemirroring/reconciler.go

```diff
@@ -0,0 +1,405 @@
+
+	// Determine changes necessary for each group of slices by port map.
+	for portKey, desiredEndpoints := range desiredEndpointsByKey {
+		numEndpoints := len(desiredEndpoints)
+		pmSlicesToCreate, pmSlicesToUpdate, pmSlicesToDelete, added, removed := r.reconcileByPortMapping(
+			endpoints, existingSlicesByKey[portKey], desiredEndpoints, desiredPortsByKey[portKey], portKey.addressType())
+
+		totalAdded += added
+		totalRemoved += removed
+
+		epMetrics.Set(endpointutil.PortMapKey(portKey), mirroringmetrics.EfficiencyInfo{
+			Endpoints: numEndpoints,
+			Slices:    len(existingSlicesByKey[portKey]) + len(pmSlicesToCreate) - len(pmSlicesToDelete),
+		})
+
+		if len(pmSlicesToCreate) > 0 {
+			slicesToCreate = append(slicesToCreate, pmSlicesToCreate...)
+		}
+		if len(pmSlicesToUpdate) > 0 {
+			slicesToUpdate = append(slicesToUpdate, pmSlicesToUpdate...)
+		}
+		if len(pmSlicesToDelete) > 0 {
+			slicesToDelete = append(slicesToDelete, pmSlicesToDelete...)
+		}
+	}
+
+	// If there are unique sets of ports that are no longer desired, mark
+	// the corresponding endpoint slices for deletion.
+	for portKey, existingSlices := range existingSlicesByKey {
+		if _, ok := desiredEndpointsByKey[portKey]; !ok {
+			for _, existingSlice := range existingSlices {
+				slicesToDelete = append(slicesToDelete, existingSlice)
+			}
+		}
+	}
+
+	mirroringmetrics.EndpointsAddedPerSync.WithLabelValues().Observe(float64(totalAdded))
+	mirroringmetrics.EndpointsRemovedPerSync.WithLabelValues().Observe(float64(totalRemoved))
+
+	endpointsNN := types.NamespacedName{Name: endpoints.Name, Namespace: endpoints.Namespace}
+	r.metricsCache.UpdateEndpointPortCache(endpointsNN, epMetrics)
+
+	return r.finalize(endpoints, slicesToCreate, slicesToUpdate, slicesToDelete)
+}
+
+// reconcileByPortMapping compares the endpoints found in existing slices with
+// the list of desired endpoints and returns lists of slices to create, update,
+// and delete. The logic is split up into several main steps:
+// 1. Iterate through existing slices, delete endpoints that are no longer
+//    desired and update matching endpoints that have changed.
+// 2. Iterate through slices that have been modified in 1 and fill them up with
+//    any remaining desired endpoints.
+// 3. If there still desired endpoints left, try to fit them into a previously
+//    unchanged slice and/or create new ones.
+func (r *reconciler) reconcileByPortMapping(
+	endpoints *corev1.Endpoints,
+	existingSlices []*discovery.EndpointSlice,
+	desiredSet endpointSet,
+	endpointPorts []discovery.EndpointPort,
+	addressType discovery.AddressType,
+) ([]*discovery.EndpointSlice, []*discovery.EndpointSlice, []*discovery.EndpointSlice, int, int) {
+	slicesByName := map[string]*discovery.EndpointSlice{}
+	sliceNamesUnchanged := sets.String{}
+	sliceNamesToUpdate := sets.String{}
+	sliceNamesToDelete := sets.String{}
+	numRemoved := 0
+
+	// 1. Iterate through existing slices to delete endpoints no longer desired
```

Comment:
It feels like each discrete step should be broken up into methods on a helper object (like the desired struct example above),
then it will be easier to understand

Reply:
```
type sliceInfo struct {
  eps *EndpointSlice
  originalLen int
  updated bool
}

func (si*sliceInfo) update(endpoints) {
  if len(endpoints) == 0 { .. mark for delete }
  if ...
}

type slices map[/*slice.Name*/string]*sliceinfo

all := slices{}
for _, eps := range existingSlices {
  all[s.Name] = &sliceInfo{eps, len(eps.Endpoints), false}
}

for _, si := range all {
  newEps, changed := computeEPSUpdate(all, ...)
  if updated || len(...) { si.update(newEps) }
}

// etc...
```

Reply:
I ended up trying a few ways to clean this up and it all ended up being messier than necessary. Instead I've gone back to my original plan which was to have a 1:1 mapping between EndpointSubset and EndpointSlice and remove the vast majority of this logic. Since we were already limiting the endpoints mirrored per subset to 1000, those can call fit in a single EndpointSlice. I can't imagine custom endpoints being a common use case beyond that, and if they are, it's likely time to switch to using the EndpointSlice API. With that said, I know this is a bit of a controversial change, so I've added this as an additional commit for now so it can easily be rolled back.

Reply:
I've gone all in with this approach, I think it's cleaner and simpler. Initially I'd hoped that we could share more logic between this and the EndpointSlice controller, but it doesn't really make sense to add complexity here for the sake of uniformity with the other controller. I've integrated this approach into the original set of commits.